pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -854,7 +854,17 @@ private void sendFlowPermitsToBroker(ClientCnx cnx, int numMessages) {
                 log.debug("[{}] [{}] Adding {} additional permits", topic, subscription, numMessages);
             }
 
-            cnx.ctx().writeAndFlush(Commands.newFlow(consumerId, numMessages), cnx.ctx().voidPromise());
+            cnx.ctx().writeAndFlush(Commands.newFlow(consumerId, numMessages))
+                    .addListener(writeFuture -> {
+                        if (log.isDebugEnabled()) {
```

Comment:
in case of not "debug enabled"...can we keep using the voidPromise ?
we are not in an hotpath but we can save a little resources.

or is it worthless ?

Reply:
I agree. I think it'd be better to only add the listener if debug logging is enabled.